app/frontend/Pages/join.tsx

```diff
@@ -16,52 +18,44 @@ export default function Join() {
     }
   }, [searchParams])
 
-  const handleSubmit = async (e) => {
-    e.preventDefault()
+  const id = useId();
+  const { post, isLoading, error, setError } = usePost<ExperienceCreateResponse>({
+    url: '/api/experiences/join',
+  });
+
+  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    const formData = getFormData<{ code?: string }>(e.currentTarget);
+    const code = formData.code;
 
-    if (!code.trim()) {
-      setError('Please enter a code')
-      return
+    if (!code || code.trim() === '') {
+      setError('Please enter a code');
+      return;
     }
 
-    setIsLoading(true)
-    setError('')
-
-    try {
-      const response = await fetch('/api/experiences/join', {
-        method: 'POST',
-        headers: {
-          'Content-Type': 'application/json',
-        },
-        body: JSON.stringify({
-          code: code.trim()
-        }),
-      })
-
-      const data = await response.json()
-
-      if (response.ok) {
-        if (data.status === 'registered') {
-          // User is already registered - store JWT and redirect to experience
-          qaLogger(`User already registrated, redirecting to: ${data.url}`)
-
-          localStorage.setItem('experience_jwt', data.jwt)
-          window.location.href = data.url
-        } else if (data.status === 'needs_registration') {
-          // User needs to register - redirect to registration page
-          qaLogger(`User needs registration, redirecting to: ${data.url}`)
-
-          console.log("User needs registration")
-          window.location.href = data.url
-        }
-      } else {
-        setError(data.error || 'Failed to join experience')
+    const response = await post(
+      JSON.stringify({
+        code: code.trim(),
+      }),
+    );
+
+    if (response.error === undefined) {
+      if (response.status === 'registered') {
+        // User is already registered - store JWT and redirect to experience
+        qaLogger(`User already registrated, redirecting to: ${response.url}`)
+
+        localStorage.setItem('experience_jwt', response.jwt)
+        window.location.href = response.url
+      } else if (response.status === 'needs_registration') {
+        // User needs to register - redirect to registration page
+        qaLogger(`User needs registration, redirecting to: ${response.url}`)
+
+        console.log("User needs registration")
+        window.location.href = response.url
       }
-    } catch (err) {
-      setError('Connection error. Please try again.')
-      console.error('Join experience error:', err)
-    } finally {
-      setIsLoading(false)
+    } else {
+      setError(response.error || 'Failed to join experience')
     }
```

Comment:
nit: I prefer early returns and switch statements for readability

```suggestion
    if (response.error != null) {
      setError(response.error || 'Failed to join experience')
    }
    
    switch (response.status) {
      case 'registered':
         // User is already registered - store JWT and redirect to experience
        qaLogger(`User already registrated, redirecting to: ${response.url}`);

        localStorage.setItem('experience_jwt', response.jwt);
        window.location.href = response.url;
        break;
      case 'needs_registration':
        // User needs to register - redirect to registration page
        qaLogger(`User needs registration, redirecting to: ${response.url}`);

        console.log("User needs registration");
        window.location.href = response.url;
        break;
      default:
        console.log(`Unknown response status encountered: ${response.status}`);
        break;
    }
```